File: nodeJs/tic-tac-toe-board-game/testCase/gameTest.js
let assert = require('assert');
let Game = require('../game');
let Player = require('../player');
let mark = require('../mark');
let expect = require('chai').expect;

let player1 = new Player();
player1.name = 'dr';
player1.mark = mark.O;
let player2 = new Player();
player2.name = 'mk';
player2.mark = mark.X;
describe('mocha game test', () => {
    it('should trow an nullPointerException if create Game object and pass null', () => {
        expect(() => new Game(null, null)).to.throw();
    });
    it('should trow an allReadyMarkException if you try to set mark on not empty cell', () => {
        let game = new Game(player1, player2);
        game.startGame();
        game.setMarkInCellOfBoard(0);
        expect(() => game.setMarkInCellOfBoard(0)).to.throw("cell is allready fill");
    });
    it('should match is draw if board in all cell fill', () => {
        let game = new Game(player1, player2);
        game.startGame();
        game.setMarkInCellOfBoard(0);
        game.setMarkInCellOfBoard(1);
        game.setMarkInCellOfBoard(2);
        game.setMarkInCellOfBoard(5);
        game.setMarkInCellOfBoard(3);
        game.setMarkInCellOfBoard(6);
        game.setMarkInCellOfBoard(4);
        game.setMarkInCellOfBoard(8);
        game.setMarkInCellOfBoard(7);
        assert.equal(game.getResult(), 'Draw');
    });
    it('should match is in_progress if board in all cell not fill', () => {
        let game = new Game(player1, player2);
        game.startGame();
        game.setMarkInCellOfBoard(0);
        game.setMarkInCellOfBoard(1);
        game.setMarkInCellOfBoard(2);
        game.setMarkInCellOfBoard(5);
        game.setMarkInCellOfBoard(3);
        assert.equal(game.getResult(), 'In_Progress');
    });
    it('should match result dr is win if dr pattern complate', () => {
        let game = new Game(player1, player2);
        game.startGame();
        game.setMarkInCellOfBoard(0);
        game.setMarkInCellOfBoard(1);
        game.setMarkInCellOfBoard(3);
        game.setMarkInCellOfBoard(4);
        game.setMarkInCellOfBoard(6);
        assert.equal(game.getResult(), 'dr is Win');
    });
    it('should dr(first player) turn over after current player is mk(second player)', () => {
        let game = new Game(player1, player2);
        game.startGame();
        game.setMarkInCellOfBoard(0);
        assert.equal(game.getCurrentPlayer().name, 'mk');
    });
});

File: OOP/tic-tac-toe-board-game-app/src/com/techlab/test/cases/GameTest.java
package com.techlab.test.cases;

import static org.junit.Assert.*;

import org.junit.Test;

import com.techlab.Exception.AllReadyMarkCellException;
import com.techlab.board.Mark;
import com.techlab.game.Game;
import com.techlab.player.Player;

public class GameTest {
	Player player1 = new Player("Darshan", Mark.O);
	Player player2 = new Player("Vivek", Mark.X);
	Game game = new Game(player1, player2);

	@Test(expected = NullPointerException.class)
	public void testGameConstractorPassingValueNull_PassTwoPlayerObject_ThrowNullPointerException() {
		game = new Game(null, null);
	}

	@Test
	public void testGameConstractorPassingValueNotNull_PassTwoPlayerObject_ShouldBeNotThrowNullPointerException() {
		game = new Game(player1, player2);
	}

	@Test(expected = AllReadyMarkCellException.class)
	public void testSetMarkInCellOfBoard_AllReadyCellFillAndPlayerSecondTimePutMark_ThrowAllReadyMarkCellException()
			throws AllReadyMarkCellException {
		game.startGame();
		game.setMarkInCellOfBoard(0);
		game.setMarkInCellOfBoard(0);

	}

	@Test
	public void testSetMarkInCellOfBoardAndGetResultPassPositionOfCell_BoardInAllCellFillAnyPlayerNotWin_ResultMatchIsDraw()
			throws AllReadyMarkCellException {
		game.startGame();
		game.setMarkInCellOfBoard(0);
		game.setMarkInCellOfBoard(1);
		game.setMarkInCellOfBoard(2);
		game.setMarkInCellOfBoard(5);
		game.setMarkInCellOfBoard(3);
		game.setMarkInCellOfBoard(6);
		game.setMarkInCellOfBoard(4);
		game.setMarkInCellOfBoard(8);
		game.setMarkInCellOfBoard(7);

		String actual = game.getResult();
		String expected = "Draw";
		assertEquals(expected, actual);
	}

	@Test
	public void testSetMarkInCellOfBoardAndGetResultPassPositionOfCell_BoardInAllCellNotFillAnyPlayerNotWin_ResultMatchIsInProgress()
			throws AllReadyMarkCellException {
		game.startGame();
		game.setMarkInCellOfBoard(0);
		game.setMarkInCellOfBoard(1);
		game.setMarkInCellOfBoard(2);
		game.setMarkInCellOfBoard(5);
		game.setMarkInCellOfBoard(3);
		game.setMarkInCellOfBoard(6);
		game.setMarkInCellOfBoard(4);
		String actual = game.getResult();
		String expected = "In_Progress";
		assertEquals(expected, actual);
	}

	@Test
	public void testSetMarkInCellOfBoardAndGetResultPassPositionOfCell_DarshanPatternComplete_ResultDarshanIsWin()
			throws AllReadyMarkCellException {
		game.startGame();
		game.setMarkInCellOfBoard(0);
		game.setMarkInCellOfBoard(1);
		game.setMarkInCellOfBoard(3);
		game.setMarkInCellOfBoard(4);
		game.setMarkInCellOfBoard(6);

		String actual = game.getResult();
		String expected = "Darshan is Win";
		assertEquals(expected, actual);
	}

	@Test
	public void testCurrentPlayer_FirstDarshanTurnOver_ResultVivekTurn() throws AllReadyMarkCellException {
		game.startGame();
		game.setMarkInCellOfBoard(0);

		String actual = game.getCurrentPlayer().getName();
		String expected = "Vivek";
		assertEquals(expected, actual);
	}

}
